cpp/include/cudf/dfa/superstate.hpp

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <type_traits>
+
+namespace cudf {
+namespace dfa {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief specialization helper class for superstate used to minimize storage requirement.
+ *
+ * @note provides no specialization for:
+ * N =  0 - not a state
+ * N =  1 - better represented by the underlying type
+ * N > 16 - better represented as multiple indivual states
+ *
+ * @tparam N number of states the superstate tracks
+ */
+template <uint8_t N, typename Enable = void>
+struct superstate_policy {
+};
```

Comment:
Something like this feels like it better conveys where these numbers come from. Plus it's more compact. 
```
template <uint8_t Bits, typename Enable = void>
struct rep{};
template<uint8_t Bits> 
struct rep <Bits, std::enable_if_t< 0  < Bits and Bits <= 8  > > { using type = uint8_t; };
template<uint8_t Bits> 
struct rep <Bits, std::enable_if_t< 8  < Bits and Bits <= 16 > > { using type = uint16_t; };
template<uint8_t Bits> 
struct rep <Bits, std::enable_if_t< 16 < Bits and Bits <= 32 > > { using type = uint32_t; };
template<uint8_t Bits>
struct rep <Bits, std::enable_if_t< 32 < Bits and Bits <= 64 > > { using type = uint64_t; };

template <uint8_t N>
struct superstate_policy {
    static_assert( N > 1 and N <= 16, "");
    static constexpr uint8_t BITS = std::ceil(std::log2(N));
    static constexpr uint8_t MASK = (1 << BITS) - 1;
    using Data = typename rep<N*BITS>::type;
};
```